src/ui/public/state_management/state_storage/hashed_item_store.js

```diff
@@ -0,0 +1,174 @@
+   */
+  constructor(sessionStorage) {
+    this._sessionStorage = sessionStorage;
+
+    // Store indexed items in descending order by touched (oldest first, newest last). We'll use
+    // this to remove older items when we run out of storage space.
+    this._indexedItems = [];
+    // Associate item hashes with the corresponding indexed items. We'll use this to quickly look
+    // up an item and update its touched date when it reoccurs in the browser history.
+    this._hashToIndexedItemMap = {};
+
+    // Build index from the persisted index. This happens when we re-open a closed tab.
+    const persistedItemIndex = this._sessionStorage.getItem(HashedItemStore.PERSISTED_INDEX_KEY);
+
+    if (persistedItemIndex) {
+      this._hashToIndexedItemMap = JSON.parse(persistedItemIndex) || {};
+      this._indexedItems = values(this._hashToIndexedItemMap);
+
+      // Order items by touched date (oldest first, newest last).
+      this._indexedItems = sortBy(this._indexedItems, 'touched');
+    }
+  }
+
+  setItem(hash, item) {
+    const isItemPersisted = this._persistItem(hash, item);
+
+    if (isItemPersisted) {
+      this._touchHash(hash);
+    }
+
+    return isItemPersisted;
+  }
+
+  getItem(hash) {
+    const item = this._sessionStorage.getItem(hash);
+
+    if (item !== null) {
+      this._touchHash(hash);
+    }
+
+    return item;
+  }
+
+  _persistItem(hash, item) {
+    try {
+      this._sessionStorage.setItem(hash, item);
+      return true;
+    } catch (e) {
+      // If there was an error then we need to make some space for the item.
+      if (this._indexedItems.length === 0) {
+        // If there's nothing left to remove, then we've run out of space and we're trying to
+        // persist too large an item.
+        return false;
+      }
+
+      // We need to try to make some space for the item by removing older items (i.e. items that
+      // haven't been accessed recently).
+      this._removeOldestItem();
+
+      // Try to persist again.
+      return this._persistItem(hash, item);
+    }
+  }
+
+  _removeOldestItem() {
+    const oldestIndexedItem = this._indexedItems.shift();
+
+    // Remove oldest item from index.
+    delete this._hashToIndexedItemMap[oldestIndexedItem.hash];
+
+    // Remove oldest item from storage.
+    this._sessionStorage.removeItem(oldestIndexedItem.hash);
+  }
+
+  _touchHash(hash) {
+    // Touching a hash indicates that it's been used recently, so it won't be the first in line
+    // when we remove items to free up storage space.
+    if (this._hashToIndexedItemMap[hash]) {
+      const indexedItem = this._hashToIndexedItemMap[hash];
+
+      // If item is already indexed, update the touched date.
+      indexedItem.touched = Date.now();
+
+      // Since the items are already sorted by touched and we're only changing one item, we can
+      // avoid a "costly" sort by just moving it to the end of the array.
+      const index = this._indexedItems.indexOf(indexedItem);
```

Comment:
If we're trying to make this efficient by avoiding sorting, I would also consider just swapping the selected item with the last element, instead of the `splice` and `push`, which are not all that efficient memory wise. To me it would be clearer to, but that's just a personal opinion ;)